virtcontainers/kata_agent.go

```diff
@@ -1149,17 +1150,25 @@ func (k *kataAgent) constraintGRPCSpec(grpcSpec *grpc.Spec, passSeccomp bool) {
 	}
 	grpcSpec.Linux.Namespaces = tmpNamespaces
 
-	// VFIO char device shouldn't not appear in the guest,
-	// the device driver should handle it and determinate its group.
-	var linuxDevices []grpc.LinuxDevice
-	for _, dev := range grpcSpec.Linux.Devices {
-		if dev.Type == "c" && strings.HasPrefix(dev.Path, vfioPath) {
-			k.Logger().WithField("vfio-dev", dev.Path).Debug("removing vfio device from grpcSpec")
-			continue
+	// If we're in guest-native-driver mode for VFIO devices, we
+	// need to remove the VFIO char devices from the container
+	// spec: the guest device driver will bind the device and
+	// create its own (different) devices.
+	// When we're not in guest native driver mode, we need to
+	// leave them in, so the agent can use the information here
+	// and elsewhere to properly bind drivers and create device
+	// node in the container.
+	if !vfioInGuest {
+		var linuxDevices []grpc.LinuxDevice
+		for _, dev := range grpcSpec.Linux.Devices {
+			if dev.Type == "c" && strings.HasPrefix(dev.Path, vfioPath) {
+				k.Logger().WithField("vfio-dev", dev.Path).Debug("removing vfio device from grpcSpec")
```

Comment:
Why are you removing the vfio devices here? I though you would simply replace them in the agent.

Seeing your agent PR after the fact, will need to look at that to see how you are handling this.

Reply:
There was already code to remove the devices here, I'm just making it conditional on the configuration.